In [1]:
# env setting
import sys
sys.path.append("../src")
sys.path.append("../models")

import numpy as np
import pandas as pd
from numpy.random import normal
from sklearn.ensemble import GradientBoostingClassifier, RandomForestClassifier
from sklearn.preprocessing import TargetEncoder
from sklearn.metrics import f1_score
from sklearn.model_selection import train_test_split, cross_val_score, StratifiedKFold
from sklearn.pipeline import Pipeline
from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import SimpleImputer, IterativeImputer, KNNImputer
from pycaret.classification import *
from tqdm import tqdm
from tqdm.notebook import tqdm as tqdm_notebook
tqdm_notebook.get_lock().locks = []

#custom function
import config
import helpers

# V2. Code
* Target Encoding으로 진행 (모델 고정 후 결과를 비교해보기)
* Missing Value를 예측값으로 Impute 하기
* 추가적인 변수 생성 해보기

In [38]:
(train, test, origin, submission) = helpers.data_loader()
train.head(5)

id surgery    age  hospital_number  rectal_temp  pulse  respiratory_rate  \
0   0     yes  adult           530001         38.1  132.0              24.0   
1   1     yes  adult           533836         37.5   88.0              12.0   
2   2     yes  adult           529812         38.3  120.0              28.0   
3   3     yes  adult          5262541         37.1   72.0              30.0   
4   4      no  adult          5299629         38.0   52.0              48.0   

  temp_of_extremities peripheral_pulse mucous_membrane  ...  \
0                cool          reduced   dark_cyanotic  ...   
1                cool           normal   pale_cyanotic  ...   
2                cool          reduced       pale_pink  ...   
3                cold          reduced       pale_pink  ...   
4              normal           normal     normal_pink  ...   

  packed_cell_volume total_protein abdomo_appearance abdomo_protein  \
0               57.0           8.5     serosanguious            3.4   
1               33.0          64.0     serosanguious            2.0   
2               37.0           6.4     serosanguious            3.4   
3               53.0           7.0            cloudy            3.9   
4               47.0           7.3            cloudy            2.6   

  surgical_lesion lesion_1  lesion_2 lesion_3 cp_data     outcome  
0             yes     2209         0        0      no        died  
1             yes     2208         0        0      no  euthanized  
2             yes     5124         0        0      no       lived  
3             yes     2208         0        0     yes       lived  
4              no        0         0        0     yes       lived  

[5 rows x 29 columns]

In [39]:
train = train.replace('None', np.NaN)
test = test.replace('None', np.NaN)
origin = origin.replace('None', np.NaN)

categorical_features = config.CATEGORICAL_FEATURES
target = 'outcome'
numerical_features = list(set(train.columns) - set(categorical_features) - set(config.USELESS_FEATURES) - set([target]))

In [40]:
print(f"TRAIN SHAPE : {train.shape}")
print(f"TEST SHAPE : {test.shape}")
print(f"ORIGINAL SHAPE : {origin.shape}")

TRAIN SHAPE : (1235, 29)
TEST SHAPE : (824, 28)
ORIGINAL SHAPE : (299, 28)


In [41]:
train['outcome'] = train['outcome'].map({
    'died' : 0,
    'euthanized' : 1,
    'lived' : 2
})

## Concat Data

In [14]:
def encode(y, target_map):
    y = np.array(y)
    encoded_y = [target_map[f] for f in y]
    return encoded_y

In [15]:
# concat train, test, origin
# first, concat train & origin
train = pd.concat(
    [train, origin], ignore_index=True
)
train = train.drop_duplicates().drop(['id'], axis=1)
test = test.drop(['id'], axis=1)


target_map = {
    'died' : 0,
    'euthanized' : 1,
    'lived' : 2
}

train['outcome'] = encode(train.outcome, target_map)

## Missing Values
train 데이터를 helpers.preprocessing output으로 고정시켜 Impute 실험을 총 4개로 진행. F1 score를 비교하였음.
1. SimpleImputer (mode, mean)
2. BeysianRidge Imputer
3. RandomForest Imputer
4. KNN Imputer

결과 :  
Train Set에 대해선 2, 3, 4 모두 0.73 정도로 약간의 차이만 있었지만  
Valid Set에 대해선 KNN Imputer가 가장 높은 <font color='orange'>F1 Score(0.702)</font>를 기록하여 KNN Imputer로 채택해보았음.

문제점:  
Hyperparameter세팅을 하지 않고 진행하였음.  
Target Encoding으로 Preprocess를 진행하면 결과가 또 달라질 수 있다고 생각됨.

In [7]:
print("Train NaN : \n", train.isna().sum())
print("Test Nan : \n", test.isna().sum())

Train NaN : 
 surgery                    0
age                        0
hospital_number            0
rectal_temp               60
pulse                     24
respiratory_rate          58
temp_of_extremities       95
peripheral_pulse         129
mucous_membrane           68
capillary_refill_time     38
pain                      99
peristalsis               64
abdominal_distention      79
nasogastric_tube         184
nasogastric_reflux       127
nasogastric_reflux_ph    246
rectal_exam_feces        292
abdomen                  331
packed_cell_volume        29
total_protein             33
abdomo_appearance        213
abdomo_protein           198
surgical_lesion            0
lesion_1                   0
lesion_2                   0
lesion_3                   0
cp_data                    0
outcome                    0
dtype: int64
Test Nan : 
 surgery                    0
age                        0
hospital_number            0
rectal_temp                0
pulse                      0
res

In [36]:
from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import SimpleImputer, IterativeImputer, KNNImputer
from sklearn.metrics import f1_score
from sklearn.model_selection import StratifiedKFold
from sklearn.ensemble import RandomForestRegressor

le_cols = ['surgery', 'age', 'surgical_lesion', 'cp_data']
ohe_cols = ['mucous_membrane']

train_preprocessing = helpers.preprocessing(train.copy(), le_cols, ohe_cols)

train_encoding_v1 = train.copy().drop(['outcome'], axis=1) # Mode, Mean
train_encoding_v2 = train_preprocessing.copy().drop(['outcome'], axis=1) # Baysian Ridge
train_encoding_v3 = train_preprocessing.copy().drop(['outcome'], axis=1) # Tree
train_encoding_v4 = train_preprocessing.copy().drop(['outcome'], axis=1) # KNN

train_encoding_v1[categorical_features] = (SimpleImputer(strategy="most_frequent").set_output(transform="pandas")
                                           .fit_transform(train_encoding_v1[categorical_features]))

train_encoding_v1[numerical_features] = (SimpleImputer(strategy="mean").set_output(transform="pandas")
                                         .fit_transform(train_encoding_v1[numerical_features]))

train_encoding_v2 = (IterativeImputer(random_state=42).set_output(transform="pandas")
                    .fit_transform(train_encoding_v2, train.outcome))

train_encoding_v3 = (IterativeImputer(estimator= RandomForestRegressor(random_state=42), random_state=42).set_output(transform="pandas")
                                     .fit_transform(train_encoding_v3, train.outcome))

train_encoding_v4 = (KNNImputer(n_neighbors=5).set_output(transform="pandas")
                    .fit_transform(train_encoding_v4))


In [52]:
from sklearn.ensemble import RandomForestClassifier

train_experiments = [helpers.preprocessing(train_encoding_v1, le_cols, ohe_cols), train_encoding_v2, train_encoding_v3, train_encoding_v4]
target = train.outcome.map({
    'died' : 0,
    'euthanized' : 1,
    'lived' : 2
})

rf = RandomForestClassifier(max_depth=6, random_state=42)
folds = StratifiedKFold(n_splits=5)

f1_score_fold = {}
for ind, train_ in enumerate(train_experiments):
    print(f'imputation version {ind + 1}')
    scores_trn, scores_val = [], []
    for fold_, (train_ind, val_ind) in enumerate(folds.split(train_, target)):
        trn_data, val_data = train_.iloc[train_ind], train_.iloc[val_ind]
        y_trn, y_val = target.iloc[train_ind], target.iloc[val_ind]
        
        rf = RandomForestClassifier(max_depth=4, random_state=42)
        rf.fit(trn_data, y_trn)
        y_hat = rf.predict(trn_data)
        y_hat_val = rf.predict(val_data)
        
        train_score = f1_score(y_trn, y_hat, average = 'micro')
        val_score = f1_score(y_val, y_hat_val, average = 'micro')
        
        scores_trn.append(train_score)
        scores_val.append(val_score)
        
    print(f"        [Train]: {np.mean(scores_trn)}")
    print(f"        [Valid]: {np.mean(scores_val)}")


imputation version 1
        [Train]: 0.7310951474558587
        [Valid]: 0.6910029592727428
imputation version 2
        [Train]: 0.7337028689449385
        [Valid]: 0.7001447701773434
imputation version 3
        [Train]: 0.7333770033104232
        [Valid]: 0.6929722594792531
imputation version 4
        [Train]: 0.7333774015169012
        [Valid]: 0.7020991675714803


## Target Encoding with KNN Imputation
https://www.kaggle.com/code/vprokopev/mean-likelihood-encodings-a-comprehensive-study/notebook -> 공부하기 좋은 링크  

위 링크에서 먼저 label Encoding, Frequency Encoding을 사용해보라고 하여서 Frequency Encoding과 비교 
Encoding 후 KNN Imputer로 결측값 채워넣고 진행.

결과:
Freq, Target 방법이 n_neighbors = 11에서 `CV F1 Score (0.72425)`의 최고점을 기록하였음.  
따라서 둘 중에 하나만 선택해도 될 것 같음.

In [21]:
from numpy.random import normal
from sklearn.ensemble import GradientBoostingClassifier, RandomForestClassifier
from sklearn.preprocessing import TargetEncoder
from sklearn.metrics import f1_score
from sklearn.model_selection import train_test_split, cross_val_score, StratifiedKFold
from sklearn.pipeline import Pipeline
from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import SimpleImputer, IterativeImputer, KNNImputer


    
def freq_encode(train_data, columns):
    encoded_cols = []
    n_samples = train_data.shape[0]
    for col in columns:
        freqs_cat = train_data.groupby(col)[col].count() / n_samples
        encoded_col_train = train_data[col].map(freqs_cat)
        encoded_cols.append(pd.DataFrame({'freq_'+ col:encoded_col_train}))
    return pd.concat(encoded_cols, axis=1)

def target_encode(train_data, target, columns, alpha=0.5):
    n_rows = train_data.shape[0]
    global_mean = train_data[target].mean()
    encoded_cols = []
    
    for col in columns:
        target_mean = train_data.groupby(col)[target].mean()
        smoothing = (target_mean * n_rows + global_mean * alpha) / (n_rows + alpha)
        encoded_col_train = train_data[col].map(smoothing)
        #encoded_col_test = test_data[col].map(smoothing)
        encoded_cols.append(pd.DataFrame({'target_' + col:encoded_col_train}))
        
    return pd.concat(encoded_cols, axis=1)
    


In [51]:
le_cols = ['surgery', 'age', 'surgical_lesion', 'cp_data']
ohe_cols = ['mucous_membrane']

rf = RandomForestClassifier(max_depth=6, random_state=42)
folds = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

train_experiments = []
for _ in range(3):
    train_experiments.append(train.copy())

y = train.copy().outcome

freq_cat = freq_encode(train_experiments[0],  categorical_features)
train_experiments[0] = pd.concat([train_experiments[0].drop(categorical_features, axis=1), freq_cat], axis=1).drop(['outcome'], axis=1) # concat with numerics

train_experiments[1] = helpers.preprocessing(train_experiments[1], le_cols, ohe_cols).drop(['outcome'], axis=1)

targ_cat = target_encode(train_experiments[2], 'outcome', categorical_features, alpha=0.5)
train_experiments[2] = pd.concat([train_experiments[2].drop(categorical_features, axis=1), freq_cat], axis=1).drop(['outcome'], axis=1) # concat with numerics

# freq, target은 n_neighbors=11일때 잘 작동

for setup_x in train_experiments:
    setup_x = (KNNImputer(n_neighbors=11).set_output(transform="pandas")
               .fit_transform(setup_x))
    
    _scores = cross_val_score(rf, setup_x, y,
                                cv=folds, scoring='f1_micro',
                                n_jobs=-1)
    
    print(f"CV Average Score : {round(np.mean(_scores), 5)}")
    print(f"    {_scores}")


CV Average Score : 0.72425
    [0.72312704 0.71335505 0.70684039 0.75895765 0.71895425]
CV Average Score : 0.70079
    [0.70358306 0.71009772 0.65798046 0.72312704 0.70915033]
CV Average Score : 0.72425
    [0.72312704 0.71335505 0.70684039 0.75895765 0.71895425]


### KNN Imputer
KNN Imputer의 경우 n_neighbors parameter를 바꿔가며 실험해 보자. 
<font color='gray'>PS S3E22 | EDA - Preprocessing - Ensemble</font>

결과:  
n_neighbors 수가 10인 경우가 CV Score 가 7.13으로 가장 높게 나타났다.


In [138]:
def knn_imputer_test(X, y, n_search_grid:list):
    train_knn = X.copy()
    
    rf = RandomForestClassifier(max_depth=6, random_state=42)
    folds = StratifiedKFold(n_splits=5)
    
    target = y.map({
        'died' : 0,
        'euthanized' : 1,
        'lived' : 2
    })
    
    knn_expirement_data = []
    for n in n_search_grid:
        _result = (KNNImputer(n_neighbors=n).set_output(transform="pandas")
                .fit_transform(train_knn))
        
        scores = cross_val_score(rf, _result, target,
                                cv=folds, scoring='f1_micro',
                                n_jobs=-1)
        
        print(f"{n} Neighbors CV Average Score : {round(np.mean(scores), 5)}")
        print(f"    {scores}")

In [142]:
# test n_neighbors = 10
train_knn = train_preprocessing.copy().drop(['outcome'], axis=1) # KNN

# init model, folds
rf = RandomForestClassifier(max_depth=6, random_state=42)
folds = StratifiedKFold(n_splits=5)

target = train_preprocessing.copy().outcome.map({
    'died' : 0,
    'euthanized' : 1,
    'lived' : 2
})

grid = [3, 5, 7, 10, 12, 15, 20]
knn_expirement_data = []
for n in grid:
    _result = (KNNImputer(n_neighbors=n).set_output(transform="pandas")
               .fit_transform(train_knn))
    
    scores = cross_val_score(rf, _result, target,
                             cv=folds, scoring='f1_micro',
                             n_jobs=-1)
    
    print(f"{n} Neighbors CV Average Score : {round(np.mean(scores), 5)}")
    print(f"    {scores}")

3 Neighbors CV Average Score : 0.70276
    [0.66775244 0.70358306 0.70358306 0.70684039 0.73202614]
5 Neighbors CV Average Score : 0.70796
    [0.6742671  0.71009772 0.71661238 0.72312704 0.71568627]
7 Neighbors CV Average Score : 0.70471
    [0.66449511 0.70032573 0.70032573 0.73289902 0.7254902 ]
10 Neighbors CV Average Score : 0.71318
    [0.67100977 0.70358306 0.72312704 0.72964169 0.73856209]
12 Neighbors CV Average Score : 0.70797
    [0.67752443 0.69381107 0.71335505 0.72638436 0.72875817]
15 Neighbors CV Average Score : 0.70471
    [0.66775244 0.69381107 0.70684039 0.72638436 0.72875817]
20 Neighbors CV Average Score : 0.7034
    [0.65798046 0.70358306 0.71009772 0.71986971 0.7254902 ]


## Pycaret Modeling
중간 Test 결과:  
Highest Score : 0.705 -> 0.76

In [52]:
from pycaret.classification import *

s = setup(data = train, target = 'outcome', categorical_features=categorical_features,
          normalize=True, fold=5, session_id=42)

compare_models()

ExtraTreesClassifier(bootstrap=False, ccp_alpha=0.0, class_weight=None,
                     criterion='gini', max_depth=None, max_features='sqrt',
                     max_leaf_nodes=None, max_samples=None,
                     min_impurity_decrease=0.0, min_samples_leaf=1,
                     min_samples_split=2, min_weight_fraction_leaf=0.0,
                     n_estimators=100, n_jobs=-1, oob_score=False,
                     random_state=42, verbose=0, warm_start=False)

In [57]:
def target_encode(train_data, test_data, target, columns, alpha=0.5):
    n_rows = train_data.shape[0]
    global_mean = train_data[target].mean()
    encoded_cols_train = []
    encoded_cols_test = []
    
    for col in columns:
        target_mean = train_data.groupby(col)[target].mean()
        smoothing = (target_mean * n_rows + global_mean * alpha) / (n_rows + alpha)
        
        encoded_col_train = train_data[col].map(smoothing)
        encoded_col_test = test_data[col].map(smoothing)
        
        encoded_cols_train.append(pd.DataFrame({col:encoded_col_train}))
        encoded_cols_test.append(pd.DataFrame({col:encoded_col_test}))
        
    return pd.concat(encoded_cols_train, axis=1), pd.concat(encoded_cols_test, axis=1)

df_train = train.copy()
df_test = test.copy()

tar_train, tar_test = target_encode(df_train, df_test, "outcome", categorical_features, alpha=0.5)
df_train = pd.concat([df_train.drop(categorical_features, axis=1), tar_train], axis=1) # concat with numerics

df_train = (KNNImputer(n_neighbors=11).set_output(transform="pandas")
            .fit_transform(df_train))

s = setup(data = df_train, target = 'outcome', categorical_features=categorical_features,
          normalize=True, fold=5, session_id=42)

compare_models()

GradientBoostingClassifier(ccp_alpha=0.0, criterion='friedman_mse', init=None,
                           learning_rate=0.1, loss='log_loss', max_depth=3,
                           max_features=None, max_leaf_nodes=None,
                           min_impurity_decrease=0.0, min_samples_leaf=1,
                           min_samples_split=2, min_weight_fraction_leaf=0.0,
                           n_estimators=100, n_iter_no_change=None,
                           random_state=42, subsample=1.0, tol=0.0001,
                           validation_fraction=0.1, verbose=0,
                           warm_start=False)

## Feature 추가 & Feature Selection

변수들끼리 사칙연산을 통해 새 Feature들을 생성하고 top N개만 사용하여 모델링

새 변수들 생성 후 결과:  
CV F1 Score (all features) : 0.76 그러나 편차가 심해보임 (0.79 ~ 0.71)  
-> Pycaret highest F1 Score : 0.7559

In [122]:
def feature_transform(data):
    data['lesion_2'] = data['lesion_2'].apply(lambda x: 1 if x > 0 else 0)
    data["deviation_from_normal_temp"] = data["rectal_temp"].apply(lambda x: abs(x - 37.8))
    
    return data

In [123]:
df_train = train.copy()
df_test = test.copy()

df_train = feature_transform(df_train)
df_test = feature_transform(df_test)

tar_train, tar_test = target_encode(df_train, df_test, "outcome", categorical_features, alpha=0.5)
df_train = pd.concat([df_train.drop(categorical_features, axis=1), tar_train], axis=1) # concat with numerics

df_train = (KNNImputer(n_neighbors=11).set_output(transform="pandas")
            .fit_transform(df_train))

In [124]:
cols = df_train.columns.drop('outcome')
X = df_train.drop(['outcome'], axis=1)
y = df_train.outcome
for idx in range(0, len(cols) -1):
    col1 = cols[idx]
    col2 = cols[idx + 1]
    
    # 곱하기
    X[col1 + "*" + col2] = X[col1] + X[col2]
    
    # 나누기
    X[col1 + "/" + col2] = X[col1] / (X[col2] + 1e-6)
    
    # 빼기
    X[col1 + "-" + col2] = X[col1] - X[col2]
    
    # 더하기
    X[col1 + "+" + col2] = X[col1] + X[col2]

    

C:\Users\jhk32\AppData\Local\Temp\ipykernel_18796\3244077655.py:18: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  X[col1 + "+" + col2] = X[col1] + X[col2]
C:\Users\jhk32\AppData\Local\Temp\ipykernel_18796\3244077655.py:9: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  X[col1 + "*" + col2] = X[col1] + X[col2]
C:\Users\jhk32\AppData\Local\Temp\ipykernel_18796\3244077655.py:12: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider j

In [127]:
from lightgbm import LGBMClassifier

lgbm = LGBMClassifier()
lgbm.fit(X, y)

scores = cross_val_score(lgbm, X, y,
                         cv=folds, scoring='f1_micro',
                         n_jobs=-1)

print(np.mean(scores))
print(scores)

[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.001607 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 14681
[LightGBM] [Info] Number of data points in the train set: 1534, number of used features: 134
[LightGBM] [Info] Start training from score -1.147370
[LightGBM] [Info] Start training from score -1.648659
[LightGBM] [Info] Start training from score -0.712898
0.7620489238040493
[0.79478827 0.71661238 0.76221498 0.79153094 0.74509804]


In [128]:
top_cols = pd.DataFrame(lgbm.feature_importances_, index=lgbm.feature_name_).sort_values(0, ascending=False).head(60).index

In [129]:
X = X[top_cols]
X['outcome'] = y

s = setup(data=X, target='outcome', normalize=True, fold=5, session_id=42)
compare_models()

ExtraTreesClassifier(bootstrap=False, ccp_alpha=0.0, class_weight=None,
                     criterion='gini', max_depth=None, max_features='sqrt',
                     max_leaf_nodes=None, max_samples=None,
                     min_impurity_decrease=0.0, min_samples_leaf=1,
                     min_samples_split=2, min_weight_fraction_leaf=0.0,
                     n_estimators=100, n_jobs=-1, oob_score=False,
                     random_state=42, verbose=0, warm_start=False)

## Submission

최종적으로는 Feature를 추가하고 상위 60개만 사용  
결과:  
(ExtraTree) Test Score는 0.7481로 V1보다는 상승했지만 실제 제출 후 score는 0.72로 하락하였음. ?
(LGBM) Test Score는 0.7590

In [16]:
def feature_transform(data):
    data['lesion_2'] = data['lesion_2'].apply(lambda x: 1 if x > 0 else 0)
    data["deviation_from_normal_temp"] = data["rectal_temp"].apply(lambda x: abs(x - 37.8))
    
    return data

def target_encode(train_data, test_data, target, columns, alpha=0.5):
    n_rows = train_data.shape[0]
    global_mean = train_data[target].mean()
    encoded_cols_train = []
    encoded_cols_test = []
    
    for col in columns:
        target_mean = train_data.groupby(col)[target].mean()
        smoothing = (target_mean * n_rows + global_mean * alpha) / (n_rows + alpha)
        
        encoded_col_train = train_data[col].map(smoothing)
        encoded_col_test = test_data[col].map(smoothing)
        
        encoded_cols_train.append(pd.DataFrame({col:encoded_col_train}))
        encoded_cols_test.append(pd.DataFrame({col:encoded_col_test}))
        
    return pd.concat(encoded_cols_train, axis=1), pd.concat(encoded_cols_test, axis=1)

def feature_engineering(data):
    cols = data.columns
    X = data.copy()

    for idx in range(0, len(cols) -1):
        col1 = cols[idx]
        col2 = cols[idx + 1]
        
        # 곱하기
        X[col1 + "*" + col2] = X[col1] + X[col2]
        
        # 나누기
        X[col1 + "/" + col2] = X[col1] / (X[col2] + 1e-6)
        
        # 빼기
        X[col1 + "-" + col2] = X[col1] - X[col2]
        
        # 더하기
        X[col1 + "+" + col2] = X[col1] + X[col2]
    
    return X


# get train, test
df_train = train.copy()
df_test = test.copy()

# Add New Features
df_train = feature_transform(df_train)
df_test = feature_transform(df_test)

# Encoding
tar_train, tar_test = target_encode(df_train, df_test, "outcome", columns = categorical_features, alpha=0.5)
df_train = pd.concat([df_train.drop(categorical_features, axis=1), tar_train], axis=1) # concat with numerics

df_test = pd.concat([df_test.drop(categorical_features, axis=1), tar_test], axis=1)

# Split data, target
y = df_train.outcome
df_train = df_train.drop(['outcome'], axis=1)

# Imputation
knn = KNNImputer(n_neighbors=11).set_output(transform="pandas")
df_train = knn.fit_transform(df_train)

df_test = knn.transform(df_test)

df_train = feature_engineering(df_train)
df_test = feature_engineering(df_test)


from lightgbm import LGBMClassifier

lgbm = LGBMClassifier()
folds = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)
lgbm.fit(df_train, y)

scores = cross_val_score(lgbm, df_train, y,
                         cv=folds, scoring='f1_micro',
                         n_jobs=-1)

print(np.mean(scores))
print(scores)

top_cols = pd.DataFrame(lgbm.feature_importances_, index=lgbm.feature_name_).sort_values(0, ascending=False).head(60).index

[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.003045 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 14676
[LightGBM] [Info] Number of data points in the train set: 1534, number of used features: 134
[LightGBM] [Info] Start training from score -1.147370
[LightGBM] [Info] Start training from score -1.648659
[LightGBM] [Info] Start training from score -0.712898
0.7666070554171722
[0.7980456  0.71661238 0.75895765 0.81758958 0.74183007]


In [17]:
X = df_train[top_cols]
X['outcome'] = y

s = setup(data=X, target='outcome', normalize=True, fold=5, session_id=42, verbose=False)
top1 = compare_models(sort='f1')

In [30]:
model = create_model('lightgbm')
tuned_model = tune_model(model, fold=10, optimize='f1', search_library='optuna')
predict_model(tuned_model)

\
                                                 
Initiated  . . . . . . . . . . . . . . . . . .   
Status     . . . . . . . . . . . . . . . . . .   
Estimator  . . . . . . . . . . . . . . . . . .   

                                            
                                            
Initiated                         14:15:29  
Status                     Initializing CV  
Estimator  Light Gradient Boosting Machine

[I 2023-09-24 14:15:39,586] Searching the best hyperparameters using 1073 samples...
[I 2023-09-24 14:16:55,448] Finished hyperparameter search!


[LightGBM] [Warning] feature_fraction is set=0.8716926056166499, colsample_bytree=1.0 will be ignored. Current value: feature_fraction=0.8716926056166499
[LightGBM] [Warning] bagging_fraction is set=0.4206344874719692, subsample=1.0 will be ignored. Current value: bagging_fraction=0.4206344874719692
[LightGBM] [Warning] bagging_freq is set=3, subsample_freq=0 will be ignored. Current value: bagging_freq=3
[LightGBM] [Warning] feature_fraction is set=0.8716926056166499, colsample_bytree=1.0 will be ignored. Current value: feature_fraction=0.8716926056166499
[LightGBM] [Warning] bagging_fraction is set=0.4206344874719692, subsample=1.0 will be ignored. Current value: bagging_fraction=0.4206344874719692
[LightGBM] [Warning] bagging_freq is set=3, subsample_freq=0 will be ignored. Current value: bagging_freq=3


hospital_number*temp_of_extremities  packed_cell_volume/total_protein  \
601                              1.560281                          5.072463   
840                              2.597443                          7.833332   
422                              2.282359                          7.428570   
966                              2.047622                          7.230768   
1414                             1.664413                          6.266666   
...                                   ...                               ...   
1320                             2.412677                          7.258063   
856                              2.045836                          0.792683   
1442                             2.747394                          0.719101   
503                              2.960709                          7.058823   
540                              2.664088                          1.000000   

      hospital_number  deviation_from_normal_temp*hospital_number  \
601          0.562699                                    0.762699   
840          1.599861                                    3.299861   
422          0.750138                                    0.750138   
966          1.050040                                    1.550040   
1414         0.666832                                    1.266832   
...               ...                                         ...   
1320         0.857246                                    1.057246   
856          1.153852                                    2.953852   
1442         1.749812                                    1.749812   
503          1.428488                                    2.128488   
540          1.666506                                    4.666506   

      lesion_1/deviation_from_normal_temp  abdomo_protein/lesion_1  \
601                          6.999965e+03                 0.002071   
840                          3.594704e+03                 0.000638   
422                          3.205000e+09                 0.000624   
966                          4.417991e+03                 0.001856   
1414                         3.833327e+03                 0.001443   
...                                   ...                      ...   
1320                         1.999990e+03                 0.008886   
856                          2.388887e+03                 0.000465   
1442                         3.205000e+09                 0.000604   
503                          3.154281e+03                 0.001042   
540                          7.359998e+02                 0.000906   

      nasogastric_reflux_ph/packed_cell_volume  \
601                                   0.128571   
840                                   0.091489   
422                                   0.019231   
966                                   0.095745   
1414                                  0.138298   
...                                        ...   
1320                                  0.102424   
856                                   0.030769   
1442                                  0.031960   
503                                   0.072917   
540                                   0.030769   

      hospital_number/temp_of_extremities  \
601                              0.564062   
840                              1.603737   
422                              0.489575   
966                              1.052584   
1414                             0.668447   
...                                   ...   
1320                             0.551130   
856                              1.293579   
1442                             1.754052   
503                              0.932298   
540                              1.670544   

      deviation_from_normal_temp-hospital_number  \
601                                    -0.362699   
840                                     0.100139   
422                                    -0.750138   
966                                  

In [31]:
final_model = finalize_model(tuned_model)

X_test = df_test[top_cols]
prediction = predict_model(final_model, data = X_test)

decode_map = {
    0 : 'died',
    1 : 'euthanized',
    2 : 'lived'
}

sample_submission = pd.read_csv(config.SUBMISSION_FILE)
sample_submission['outcome'] = prediction['prediction_label'].map(decode_map)
sample_submission.to_csv("../output/sample_submission_V2(2nd).csv", index=False)

[LightGBM] [Warning] feature_fraction is set=0.8716926056166499, colsample_bytree=1.0 will be ignored. Current value: feature_fraction=0.8716926056166499
[LightGBM] [Warning] bagging_fraction is set=0.4206344874719692, subsample=1.0 will be ignored. Current value: bagging_fraction=0.4206344874719692
[LightGBM] [Warning] bagging_freq is set=3, subsample_freq=0 will be ignored. Current value: bagging_freq=3
[LightGBM] [Warning] feature_fraction is set=0.8716926056166499, colsample_bytree=1.0 will be ignored. Current value: feature_fraction=0.8716926056166499
[LightGBM] [Warning] bagging_fraction is set=0.4206344874719692, subsample=1.0 will be ignored. Current value: bagging_fraction=0.4206344874719692
[LightGBM] [Warning] bagging_freq is set=3, subsample_freq=0 will be ignored. Current value: bagging_freq=3


In [ ]:
# models = {
#     'et' : create_model("et", fold=5),
#     'catboost' : create_model("catboost", fold=5),
#     'gbc' : create_model("gbc", fold=5),
#     'xgboost' : create_model("xgboost", fold=5),
#     'lightgbm' : create_model("lightgbm", fold=5)
# }


# for model_name, model in models.items():
#     try:
#         models[model_name] = tune_model(model, 
#                                         optimize="F1",
#                                         fold=5,
#                                         search_library="optuna")
        
#     except:
#         models[model_name] = tune_model(model, 
#                                         optimize="F1",
#                                         fold=5,
#                                         search_library="scikit-learn")
        


In [ ]:
blended_models = blend_models(models, fold=5, optimize="f1")

In [ ]:
def fit(X, y):
        n_rows = X.shape[0]
        columns = X.columns.values
        global_mean = y.mean()
        encoded_cols_X = []
        
        for col in columns:
            target_mean = X.groupby(col)[y.name].mean()
            smoothing = (target_mean * n_rows + global_mean * alpha) / (n_rows + alpha)
            self._smoothing[col] = smoothing
            
            encoded_col_X = X[col].map(smoothing)
            
            encoded_cols_X.append(pd.DataFrame({col:encoded_col_X}))
        
        return None

[]

In [16]:
train.outcome

0             died
1       euthanized
2            lived
3            lived
4            lived
           ...    
1230         lived
1231          died
1232         lived
1233         lived
1234         lived
Name: outcome, Length: 1235, dtype: object